src/schedulers/distance-based-slot.scheduler.ts

```diff
@@ -1,38 +1,87 @@
+  distance: number
+}
+
+type GenerateParams = BaseGenerateParams<DistanceInput>
+
 /**
  * DistanceBasedSlotScheduler
  *
  * Select slots randomly between the latest known slot and a future slot.
  */
-export class DistanceBasedSlotScheduler extends SlotScheduler<TimeInput> {
-  // TODO: ensure no collision
-  public generateSchedule(
-    n: number,
-    currentSlot: number,
-    input: TimeInput
-  ): SlotSchedule {
-    // const currentSlot = Math.floor(input.currentSlot + 1);
-    const distance = Math.floor(input.distance)
-
-    if (n > distance) {
-      throw new Error('number of slots must be less than total slots')
+export class DistanceBasedSlotScheduler extends SlotScheduler<DistanceInput> {
+  /**
+   * generateSchedule
+   *
+   * Generates a schedule of `slotAmount` random slots in a range from
+   * the current slot + 1, and current slot + distance.
+   *
+   * @param {number} params.slotAmount - The number of slots to generate.
+   * @param {number} params.currentSlot - The current slot number.
+   * @param {number} params.distance - The distance (range) to sample slots from.
+   * @returns {SlotSchedule}
+   */
+  public generateSchedule(params: GenerateParams): SlotSchedule {
+    const { slotAmount, currentSlot, distance: rawDistance } = params
+
+    // TODO: Check that `currentSlot` and `distance` are integers?
+
+    const distance = Math.floor(rawDistance)
+
+    if (slotAmount > distance) {
+      throw new Error(
+        'DistanceBasedSlotScheduler: Cannot sample more slots than the available ones in the provided range.'
+      )
     }
 
-    let terminalSlot = currentSlot + 1 + distance * 2
-    const slotIds: number[] = []
+    const slotIndexes = this._pickDistinctIntegers(distance - 1, slotAmount)
+    const schedule = new SlotSchedule()
 
-    // Generate n random slot IDs between currentSlot+1 and terminalSlot
-    // ensuring multiples of 2
-    for (let i = 0; i < n; i++) {
-      const range = Math.floor((terminalSlot - currentSlot + 1) / 2)
-      const randomSlot = currentSlot + 2 + Math.floor(Math.random() * range) * 2
-      slotIds.push(randomSlot)
+    slotIndexes.forEach((slotIndex) => {
+      // Calculate `slotId`, ensuring it's a multiple of 2.
+      const slotId = currentSlot + slotIndex * 2
+      schedule.addSlot(slotId)
+    })
+
+    return schedule
+  }
+
+  /**
+   * _generateDistinctRandomIntegers
+   *
+   * Generates a prescribed amount of random integers in a
+   * specified range (from 0 to `range`).
+   *
+   * @param {number} range - The range to generate random integers from.
+   * @param {number} amount - Total amount of random integers to generate.
+   * @returns {number[]} - The generated integers.
+   */
+  private _pickDistinctIntegers(range: number, amount: number): number[] {
+    const pickedValues = new Set<number>()
+
+    while (pickedValues.size < amount) {
+      const randomNum = this._generateRandomInteger(range)
+      pickedValues.add(randomNum)
```

Comment:
Yeah, that sounds just about right. It may be overkill to use so many methods; I just think it's good to have functionality describe itself... But I'm fine with compromising a bit in this regard